src/server/tasks/assign-texters.ts

```diff
@@ -11,6 +13,221 @@ interface Texter {
+        and updated_at < $3
+    `,
+    [campaignId, assignmentIds, ignoreAfterDate]
+  );
+};
+
+interface FreeUpTextersOptions {
+  client: PoolClient;
+  campaignId: number;
+  isArchived: boolean;
+  assignmentTargets: AssignmentTarget[];
+  onProgress?: (percentComplete: number) => void | Promise<void>;
+}
+
+export const freeUpTexters = async (options: FreeUpTextersOptions) => {
+  const {
+    client,
+    campaignId,
+    isArchived,
+    assignmentTargets,
+    onProgress
+  } = options;
+  for (let index = 0; index < assignmentTargets.length; index += 1) {
+    const assignmentTarget = assignmentTargets[index];
+    const { id: assignmentId, contactsCount } = assignmentTarget;
+    await client.query(
+      `
+        with cc_ids_to_keep as (
+          select id
+          from campaign_contact
+          where
+            campaign_id = $1
+            and archived = ${isArchived}
+            and assignment_id = $2
+          order by id asc
+          limit $3
+        )
+        update campaign_contact
+        set assignment_id = null
+        where
+          campaign_id = $4
+          and archived = ${isArchived}
+          and assignment_id = $5
+          and id not in (select id from cc_ids_to_keep)
+      `,
+      [campaignId, assignmentId, contactsCount, campaignId, assignmentId]
+    );
+
+    if (onProgress && index % 10 === 0) {
+      const stagePercentCompelte = Math.floor(
+        (index / assignmentTargets.length) * 100
+      );
+      await onProgress(stagePercentCompelte);
+    }
+  }
+};
+
+interface AssignPayloadsOptions {
+  client: PoolClient;
+  campaignId: number;
+  isArchived: boolean;
+  assignmentTargets: AssignmentTarget[];
+}
+
+export const assignPayloads = async (options: AssignPayloadsOptions) => {
+  const { client, campaignId, isArchived, assignmentTargets } = options;
+
+  const assignmentIds = assignmentTargets.map(({ id }) => parseInt(id, 10));
+  const contactsCounts = assignmentTargets.map(
+    ({ contactsCount }) => contactsCount
+  );
+
+  await client.query(
+    `
+      with raw_assignments as (
+        select * from unnest($1::integer[], $2::integer[]) as t (assignment_id, desired_count)
+      ),
+      assignment_counts as (
+        select
+          assignment_id,
+          generate_series(1, desired_count - (
+            select count(*) from campaign_contact
+            where campaign_contact.assignment_id = raw_assignments.assignment_id
+          ))
+        from raw_assignments
+      ),
+      assignments_payload as (
+        select
+          row_number() over () as row,
+          assignment_id
+        from assignment_counts
+      ),
+      assignable_contacts as (
+        select
+          row_number() over () as row,
+          id as campaign_contact_id
+          from campaign_contact
+        where
+          campaign_id = $3
+          and archived = ${isArchived}
+          and assignment_id is null
+          and message_status = 'needsMessage'
```

Comment:
Should we be able to assign needs reply too?

I think we should preferably assign `needsMessage` and `needsResponse`, but using this section you should be able to assign all contacts on the campaign, right?

Also, performance wise this query would be improved with a limit on this stage in the CTE